File: computation_estimation/estimate_mkt_clear.py
import random
import math
import time
import collections
import cython
import numpy

# first generate the random numbers (200,000)

bids = []
asks = []

M = 100000

for i in xrange(M):
    bids.append(int(math.floor(random.normalvariate(1000, 20))))
    asks.append(int(math.floor(random.normalvariate(1050, 10))))

init = time.time()

# Process the data

bd = {}
ad = {}

barr = numpy.array(bids)
aarr = numpy.array(asks)

counts = numpy.bincount(barr)

counts = [(i, v) for i, v in enumerate(counts) if v]

print counts

tc = time.time()

for bid in bids:
    try:
        bd[bid] = bd[bid] + 1
    except:
        bd[bid] = 1

for ask in asks:
    try:
        ad[ask] = ad[ask] + 1
    except:
        ad[ask] = 1

t2 = time.time()

print len(bids), len(asks), len(bd.keys()), len(ad.keys())

#print "Aggregate S/D Curve: ", (t2 - init) * 1000
print "AGG SD: ", 1000*(tc - init), 1000*(t2 - tc)

t2 = time.time()

# Find the market clearing price

maxbid = int(max(bd.keys()))
minask = int(min(ad.keys()))

demand = 0
supply = 0

prevd = demand
prevs = supply

guess = 0

print "MB, MA", maxbid, minask

if maxbid > minask:

    guess = int(math.floor((maxbid + minask) / 2))

    while True:

        prevd = demand
        prevs = supply

        #guess = int(math.floor((maxbid + minask) / 2))
        print "guess", guess

        demand = 0
        supply = 0

        for price in xrange(guess, maxbid):
            try:
                demand = demand + bd[price]
            except:
                demand = demand

        for price in xrange(minask, guess):
            try:
                supply = supply + ad[price]
            except:
                supply = supply

        print "D/S:", demand, supply

        if demand == supply:
            break

        if demand > supply and prevd < prevs:
            if min(prevd, prevs) > min(demand, supply):
                demand = prevd
                supply = prevs
            break

        if demand < supply and prevd > prevs:
            if min(prevd, prevs) > min(demand, supply):
                demand = prevd
                supply = prevs
            break

        if demand > supply:
            guess = guess + 1
        elif demand < supply:
            guess = guess - 1

t3 = time.time()    

print demand, supply, guess

print "Find Market Clearing Price: ", (t3 - t2) * 1000

File: computation_estimation/estimate_mkt_clear.cpp
#include<iostream>
#include<random>
#include<vector>
#include<map>
#include<algorithm>

#include <time.h>

using namespace std;

int test(int n){

  // generate random numbers
  random_device rd;
  default_random_engine gen(rd());
  normal_distribution<double> distbid(78600, 10000);
  normal_distribution<double> distask(78505, 10000);

  vector<int> bids;
  vector<int> asks;

  for (int i = 0; i < n; i++){
    bids.push_back(floor(distbid(gen)));
    asks.push_back(floor(distask(gen)));
  }

  // Aggregate S/D into lists [price:qty]
  timespec t0;
  clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &t0);
  
  int maxbid = *max_element(bids.begin(), bids.end());
  //int minbid = *min_element(bids.begin(), bids.end());
  int maxask = *max_element(asks.begin(), asks.end());
  int minask = *min_element(asks.begin(), asks.end());
  
  vector<unsigned int> b(maxbid+1);
  vector<unsigned int> a(maxask+1);

  for (unsigned int i = 0; i < bids.size(); i++){
    b[bids[i]]++;
  }
  
  for (unsigned int i = 0; i < bids.size(); i++){
    a[asks[i]]++;
  }
  
  timespec t1, t2;
  clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &t1);

  // Find Market Clearing Prices

  int demand = -1;
  int supply = -1;
  int prevd = 0;
  int prevs = 0;

  int guess = 0;
  int prevg = 0;
  
  if (maxbid > minask){
    //guess = int(floor((maxbid + minask) / 2.0));
    //cout << "GS: \t" << guess << endl;
    guess = minask;

    while (1 == 1){

      prevd = demand;
      prevs = supply;

      if (demand == -1 || supply == -1){
	demand = 0;
	supply = 0;

	for (int i = guess; i <= maxbid; i++){
	  demand += b[i];
	}

	for (int i = minask; i <= guess; i++){
	  supply += a[i];
	}
      }
      else{
	// if price goes up
	if (prevg < guess){
	  supply += a[guess];
	  demand -= b[prevg];
	}
	// if price goes down
	else if (prevg > guess){
	  supply -= a[prevg];
	  demand += b[guess];
	}
      }

      // If the Market crosses, exit

      if (demand == supply)
	break;

      if (demand > supply && prevd < prevs){
	if (min(prevd, prevs) > min(demand, supply)){
	  demand = prevd;
	  supply = prevs;
	}
	break;
      }

      if (demand < supply && prevd > prevs){
	if (min(prevd, prevs) > min(demand, supply)){
	  demand = prevd;
	  supply = prevs;
	}
	break;
      }

      prevg = guess;

      if (demand > supply)
	guess += 1;
      else if (demand < supply)
	guess -= 1;
    }
  }
  
  clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &t2);

  // Calculate the times for each section of the computation
  
  timespec temp, tempCLEAR, tempAGG2;

  // Total
  temp.tv_nsec = t2.tv_nsec-t0.tv_nsec;
  if (temp.tv_nsec < 0){
    temp.tv_nsec = 1000000000 + temp.tv_nsec;
    temp.tv_sec = t2.tv_sec-t0.tv_sec - 1;
  }
  else{
    temp.tv_sec = t2.tv_sec-t0.tv_sec;    
  }

  // Aggregation
  tempAGG2.tv_nsec = t1.tv_nsec-t0.tv_nsec;
  if (tempAGG2.tv_nsec < 0){
    tempAGG2.tv_nsec = 1000000000 + tempAGG2.tv_nsec;
    tempAGG2.tv_sec = t1.tv_sec-t0.tv_sec - 1;
  }
  else{
    tempAGG2.tv_sec = t1.tv_sec-t0.tv_sec;    
  }


  // Clearing
  tempCLEAR.tv_nsec = t2.tv_nsec-t1.tv_nsec;
  if (tempCLEAR.tv_nsec < 0){
    tempCLEAR.tv_nsec = 1000000000 + tempCLEAR.tv_nsec;
    tempCLEAR.tv_sec = t2.tv_sec-t1.tv_sec - 1;
  }
  else{
    tempCLEAR.tv_sec = t2.tv_sec-t1.tv_sec;    
  }

  cout<<"Total Time: "<<temp.tv_sec<<"s \t \t : \t "<<temp.tv_nsec / 1000000.0<<"ms"<<endl;
  cout<<"Aggregation Time: "<<tempAGG2.tv_sec<<"s \t : \t "<<tempAGG2.tv_nsec / 1000000.0<<"ms"<<endl;
  cout<<"Clearing Time: "<<tempCLEAR.tv_sec<<"s \t : \t "<<tempCLEAR.tv_nsec / 1000000.0<<"ms"<<endl;
  //cout << "maxbid:\t"<<maxbid<<"\t minask:\t"<<minask<<endl;
  cout <<"Market Clearing Price: "<< guess << endl;
  cout << "Demand: " << demand << "\t Supply: "<<supply << endl;
  cout << endl << endl;

  return tempCLEAR.tv_nsec / 1000000.0;
}

int main(){
  //double sum = 0;
  
  for (int i = 1; i < 7; i++){
    cout << "M = " << pow(10,i) << endl;
    test(pow(10,i));
  }
  

  //test(pow(10,5));
}
